File: Ahorn/entities/customSwitchGate.jl
module VortexHelperVortexSwitchGate
using ..Ahorn, Maple

function gateFinalizer(entity)
    x, y = Ahorn.position(entity)

    width = Int(get(entity.data, "width", 8))
    height = Int(get(entity.data, "height", 8))

    entity.data["nodes"] = [(x + width, y)]
end

@mapdef Entity "VortexHelper/VortexSwitchGate" VortexSwitchGate(
			x::Integer, y::Integer,
			width::Integer=8, height::Integer=8,
			sprite::String="block",
			behavior::String="crush",
			persistent::Bool=false,
			crushDuration::Number=0.75,
			nodes::Array{Tuple{Integer, Integer}, 1}=Tuple{Integer, Integer}[])
			

const textures = ["block", "mirror", "temple", "stars"]
const behaviors = ["crush", "shatter"]

const placements = Ahorn.PlacementDict(
    "Custom Switch Gate ($(uppercasefirst(behavior))) (Vortex Helper)" => Ahorn.EntityPlacement(
        VortexSwitchGate,
		"rectangle",
		Dict{String, Any}(
			"behavior" => behavior
        ),
        gateFinalizer
    ) for behavior in behaviors
)

Ahorn.editingOptions(entity::VortexSwitchGate) = Dict{String, Any}(
    "sprite" => textures,
	"behavior" => behaviors
)

Ahorn.nodeLimits(entity::VortexSwitchGate) = 1, 1
Ahorn.minimumSize(entity::VortexSwitchGate) = 16, 16
Ahorn.resizable(entity::VortexSwitchGate) = true, true

function Ahorn.selection(entity::VortexSwitchGate)
	behavior = get(entity.data, "behavior", "crush")
    x, y = Ahorn.position(entity)
    stopX, stopY = Int.(entity.data["nodes"][1])

    width = Int(get(entity.data, "width", 8))
    height = Int(get(entity.data, "height", 8))

    return (behavior == "crush") ? [Ahorn.Rectangle(x, y, width, height), Ahorn.Rectangle(stopX, stopY, width, height)] : Ahorn.Rectangle(x, y, width, height)
end

iconResource = "objects/switchgate/icon00"

function renderGateSwitch(ctx::Ahorn.Cairo.CairoContext, x::Number, y::Number, width::Number, height::Number, sprite::String)
    iconSprite = Ahorn.getSprite(iconResource, "Gameplay")
    
    tilesWidth = div(width, 8)
    tilesHeight = div(height, 8)

    frame = "objects/switchgate/$sprite"

    for i in 2:tilesWidth - 1
        Ahorn.drawImage(ctx, frame, x + (i - 1) * 8, y, 8, 0, 8, 8)
        Ahorn.drawImage(ctx, frame, x + (i - 1) * 8, y + height - 8, 8, 16, 8, 8)
    end

    for i in 2:tilesHeight - 1
        Ahorn.drawImage(ctx, frame, x, y + (i - 1) * 8, 0, 8, 8, 8)
        Ahorn.drawImage(ctx, frame, x + width - 8, y + (i - 1) * 8, 16, 8, 8, 8)
    end

    for i in 2:tilesWidth - 1, j in 2:tilesHeight - 1
        Ahorn.drawImage(ctx, frame, x + (i - 1) * 8, y + (j - 1) * 8, 8, 8, 8, 8)
    end

    Ahorn.drawImage(ctx, frame, x, y, 0, 0, 8, 8)
    Ahorn.drawImage(ctx, frame, x + width - 8, y, 16, 0, 8, 8)
    Ahorn.drawImage(ctx, frame, x, y + height - 8, 0, 16, 8, 8)
    Ahorn.drawImage(ctx, frame, x + width - 8, y + height - 8, 16, 16, 8, 8)

    Ahorn.drawImage(ctx, iconSprite, x + div(width - iconSprite.width, 2), y + div(height - iconSprite.height, 2))
end

function Ahorn.renderSelectedAbs(ctx::Ahorn.Cairo.CairoContext, entity::VortexSwitchGate, room::Maple.Room)
	behavior = get(entity.data, "behavior", "crush")
	
	if behavior == "crush"
		sprite = get(entity.data, "sprite", "block")
		startX, startY = Int(entity.data["x"]), Int(entity.data["y"])
		stopX, stopY = Int.(entity.data["nodes"][1])
	
		width = Int(get(entity.data, "width", 32))
		height = Int(get(entity.data, "height", 32))
	
		renderGateSwitch(ctx, stopX, stopY, width, height, sprite)
		Ahorn.drawArrow(ctx, startX + width / 2, startY + height / 2, stopX + width / 2, stopY + height / 2, Ahorn.colors.selection_selected_fc, headLength=6)
	end
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::VortexSwitchGate, room::Maple.Room)
    sprite = get(entity.data, "sprite", "block")

    x = Int(get(entity.data, "x", 0))
    y = Int(get(entity.data, "y", 0))

    width = Int(get(entity.data, "width", 32))
    height = Int(get(entity.data, "height", 32))

    renderGateSwitch(ctx, x, y, width, height, sprite)
end

end

File: Loenn/entities/custom_switch_gate.lua
local drawableNinePatch = require "structs.drawable_nine_patch"
local drawableSprite = require "structs.drawable_sprite"
local utils = require "utils"

local behaviors = {
    "crush",
    "shatter"
}

local behaviorOptions = {}
for _, behavior in ipairs(behaviors) do
    behaviorOptions[utils.titleCase(behavior)] = behavior
end

local textures = {
    "block",
    "mirror",
    "temple",
    "stars"
}

local textureOptions = {}
for _, texture in ipairs(textures) do
    textureOptions[utils.titleCase(texture)] = texture
end

local customSwitchGate = {}

customSwitchGate.name = "VortexHelper/VortexSwitchGate"
customSwitchGate.depth = 0

customSwitchGate.minimumSize = {16, 16}
customSwitchGate.fieldInformation = {
    sprite = {
        options = textureOptions,
        editable = false
    },
    behavior = {
        options = behaviorOptions,
        editable = false
    },
    crushDuration = {
        minimumValue = 0.5,
        maximumValue = 2.0
    }
}

function customSwitchGate.ignoredFields(entity)
    local ignored = {"_id", "_name"}
    if entity.behavior == "shatter" then
        table.insert(ignored, "crushDuration")
    end
    return ignored
end

customSwitchGate.nodeLimits = {1, 1}

function customSwitchGate.nodeVisibility(entity)
    return (entity.behavior == "crush") and "always" or "never"
end

function customSwitchGate.nodeLineRenderType(entity)
    return (entity.behavior == "crush") and "line" or false
end

customSwitchGate.placements = {}
for i, behavior in ipairs(behaviors) do
    customSwitchGate.placements[i] = {
        name = behavior,
        data = {
            width = 16,
            height = 16,
            sprite = "block",
            behavior = behavior,
            crushDuration = 0.75
        }
    }
end

local ninePatchOptions = {
    mode = "fill",
    borderMode = "repeat",
    fillMode = "repeat"
}

local frameTexture = "objects/switchgate/%s"
local middleTexture = "objects/switchgate/icon00"

function customSwitchGate.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 24, entity.height or 24

    local blockSprite = entity.sprite or "block"
    local frame = string.format(frameTexture, blockSprite)

    local ninePatch = drawableNinePatch.fromTexture(frame, ninePatchOptions, x, y, width, height)
    local middleSprite = drawableSprite.fromTexture(middleTexture, entity)
    local sprites = ninePatch:getDrawableSprite()

    middleSprite:addPosition(math.floor(width / 2), math.floor(height / 2))
    table.insert(sprites, middleSprite)

    return sprites
end

function customSwitchGate.selection(room, entity)
    local nodes = entity.nodes or {}
    local x, y = entity.x or 0, entity.y or 0
    local nodeX, nodeY = nodes[1].x or x, nodes[1].y or y
    local width, height = entity.width or 24, entity.height or 24

    return utils.rectangle(x, y, width, height), (entity.behavior == "crush") and {utils.rectangle(nodeX, nodeY, width, height)} or nil
end

return customSwitchGate
